Add option to define parallel jobs for make and make install

martinseener opened the pull request:
### Overview

This PR adds an option to make use of multi-core systems by enabling the make and make install commands to use (by default) all available cpu cores as the parallel jobs option to speed up the compile process. The user can also customize that to any number.

martinseener commented:
Rebased the commit by removing those 2 trailing whitespaces which prevented successful builds

Ma27 commented on manifests/init.pp:
```diff
@@ -31,6 +31,7 @@
   $python_package = 'python',
+  $use_cpus       = $::processorcount,
 ) {
   validate_string($node_path)
```
a `validate_integer` to ensure type safety is missing

Ma27 commented on README.md:
````diff
@@ -131,6 +131,17 @@ __Note:__ It is not possible to install and uninstall an instance in the same ru
+```puppet
+class { 'nodejs':
+  version      => 'stable',
+  use_cpus     => 2,
````
please use appropriate alignment here (arrows should be one space longer than the longest parameter)

Ma27 commented:
Please add typedocs for the `use_cpus` parameter to the `init.pp` and `install.pp`. Would be great if you could fix the tests and add another one.

Otherwise this looks fine for me, but I need to reserve some time to check it out on my machine. Please keep in mind that this will go to `2.0` as it's a new feature and `2.0` might take some time until it's ready...

martinseener commented:
What i've done so far:
* added typedocs for init.pp and install.pp
* added validate_integer not just for init.pp but also install.pp (since you can run that define separately and so it needs to be checked there too)
* fixed the alignment in the README

I'll now try to fix the tests, so you can merge that later!

Ma27 commented on manifests/init.pp:
```diff
@@ -31,8 +34,10 @@
   $make_install   = true,
   $node_path      = '/usr/local/node/node-default/lib/node_modules',
   $python_package = 'python',
+  $use_cpus       = $::processorcount,
```
`use_cpus` might be a bit confusing as it sounds like a `bool` parameter. I suggest something like `$cpu_cores`.

Ma27 commented on spec/classes/nodejs_spec.rb:
```diff
@@ -99,4 +100,24 @@
+      .with_use_cpus(1)
+    }
+
+    it { should contain_file('/usr/local/node/node-default') \
```
this assertion can be removed as it is not relevant for your special case (it's good enough to assert against things that should have a special behavior in a test case)

Ma27 commented on spec/classes/nodejs_spec.rb:
```diff
@@ -99,4 +100,24 @@
+      .with_target('/usr/local/node/node-v6.0.0')
+    }
+
+    it { should contain_file('/etc/profile.d/nodejs.sh') }
```
same as above

Ma27 commented:
well your specs look pretty fine, I just found three minor things that aren't critical, but should be fixed before approving. I'll try to reserve some time in order to test this feature out, then we're done and the merge can be triggered.

martinseener commented:
I've renamed everything from `use_cpus` to `cpu_cores` as i think your naming suggestion is better to understand...Is still something to fix before merge?

Btw: I've already tried that out within our Puppet Development machines and it works pretty well. make ran with all 4 cores and the build was really faster than before! But sure, try it out yourself. I've tested it on Debian Jessie 8.5 w. Puppet 4.5.2 so far.

Ma27 commented:
ok, tested it on a Ubuntu Xenial with Puppet 4.5 and it looks fine, so we can merge it actually.
It should work on other system like SUSE, but the right place for that is #160 as it's currently not officially supported.

Ma27 commented:
nice work @martinseener!
